File: tyhp-lang/vscode/src/debug/proxyLifecycle.test.ts
import assert from "node:assert/strict";
import { test } from "node:test";
import { nextProxyState, proxyIsActive, proxyIsListening } from "./proxyLifecycle";

test("start then listening reaches running", () => {
    let state = nextProxyState("stopped", { type: "startRequested" });
    assert.equal(state, "starting");
    state = nextProxyState(state, { type: "listening" });
    assert.equal(state, "running");
    assert.equal(proxyIsListening(state), true);
    assert.equal(proxyIsActive(state), true);
});

test("start while already running is a no-op", () => {
    assert.equal(nextProxyState("running", { type: "startRequested" }), "running");
    assert.equal(nextProxyState("starting", { type: "startRequested" }), "starting");
});

test("start failure from starting becomes error", () => {
    assert.equal(nextProxyState("starting", { type: "startFailed" }), "error");
    assert.equal(nextProxyState("running", { type: "startFailed" }), "running");
});

test("requested stop then expected exit becomes stopped", () => {
    let state = nextProxyState("running", { type: "stopRequested" });
    assert.equal(state, "stopping");
    state = nextProxyState(state, { type: "exited", expected: true });
    assert.equal(state, "stopped");
    assert.equal(proxyIsActive(state), false);
});

test("unexpected exit from running becomes error", () => {
    assert.equal(nextProxyState("running", { type: "exited", expected: false }), "error");
    assert.equal(nextProxyState("starting", { type: "exited", expected: false }), "error");
});

test("stop from stopped stays stopped", () => {
    assert.equal(nextProxyState("stopped", { type: "stopRequested" }), "stopped");
    assert.equal(nextProxyState("stopped", { type: "exited", expected: true }), "stopped");
});

test("cleanup exit after start failure stays error", () => {
    const failed = nextProxyState("starting", { type: "startFailed" });
    assert.equal(failed, "error");
    assert.equal(nextProxyState(failed, { type: "exited", expected: true }), "error");
});

File: tyhp-lang/phpstorm/src/unitTest/kotlin/com/tyhp/lang/debug/ProxyLifecycleTest.kt
package com.tyhp.lang.debug

import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertFalse
import kotlin.test.assertTrue

class ProxyLifecycleTest {
    @Test
    fun `start then listening reaches running`() {
        var state = nextProxyState(ProxyRunState.STOPPED, ProxyLifecycleEvent.StartRequested)
        assertEquals(ProxyRunState.STARTING, state)
        state = nextProxyState(state, ProxyLifecycleEvent.Listening)
        assertEquals(ProxyRunState.RUNNING, state)
        assertTrue(proxyIsListening(state))
        assertTrue(proxyIsActive(state))
    }

    @Test
    fun `start while already running is a no-op`() {
        assertEquals(
            ProxyRunState.RUNNING,
            nextProxyState(ProxyRunState.RUNNING, ProxyLifecycleEvent.StartRequested),
        )
        assertEquals(
            ProxyRunState.STARTING,
            nextProxyState(ProxyRunState.STARTING, ProxyLifecycleEvent.StartRequested),
        )
    }

    @Test
    fun `start failure from starting becomes error`() {
        assertEquals(
            ProxyRunState.ERROR,
            nextProxyState(ProxyRunState.STARTING, ProxyLifecycleEvent.StartFailed),
        )
        assertEquals(
            ProxyRunState.RUNNING,
            nextProxyState(ProxyRunState.RUNNING, ProxyLifecycleEvent.StartFailed),
        )
    }

    @Test
    fun `requested stop then expected exit becomes stopped`() {
        var state = nextProxyState(ProxyRunState.RUNNING, ProxyLifecycleEvent.StopRequested)
        assertEquals(ProxyRunState.STOPPING, state)
        state = nextProxyState(state, ProxyLifecycleEvent.Exited(expected = true))
        assertEquals(ProxyRunState.STOPPED, state)
        assertFalse(proxyIsActive(state))
    }

    @Test
    fun `unexpected exit from running becomes error`() {
        assertEquals(
            ProxyRunState.ERROR,
            nextProxyState(ProxyRunState.RUNNING, ProxyLifecycleEvent.Exited(expected = false)),
        )
        assertEquals(
            ProxyRunState.ERROR,
            nextProxyState(ProxyRunState.STARTING, ProxyLifecycleEvent.Exited(expected = false)),
        )
    }

    @Test
    fun `stop from stopped stays stopped`() {
        assertEquals(
            ProxyRunState.STOPPED,
            nextProxyState(ProxyRunState.STOPPED, ProxyLifecycleEvent.StopRequested),
        )
        assertEquals(
            ProxyRunState.STOPPED,
            nextProxyState(ProxyRunState.STOPPED, ProxyLifecycleEvent.Exited(expected = true)),
        )
    }

    @Test
    fun `cleanup exit after start failure stays error`() {
        val failed = nextProxyState(ProxyRunState.STARTING, ProxyLifecycleEvent.StartFailed)
        assertEquals(ProxyRunState.ERROR, failed)
        assertEquals(ProxyRunState.ERROR, nextProxyState(failed, ProxyLifecycleEvent.Exited(expected = true)))
    }
}
